swarm/swap/swap.go

```diff
@@ -0,0 +1,347 @@
+func (dpo *DefaultPriceOracle) loadDefaultPriceMatrix() map[string]map[uint64]*PriceTag {
+	//setup matrix
+	priceMatrix := make(map[string]map[uint64]*PriceTag)
+
+	//define accounted messages from the stream protocol
+	streamProtocol := stream.Spec.Name
+	priceMatrix[streamProtocol] = make(map[uint64]*PriceTag)
+
+	//get indexes of accounted messages in order to populate the map
+	retrieveRequestMsgIndex, ok := stream.Spec.GetCode(stream.RetrieveRequestMsg{})
+	if !ok {
+		log.Crit("setting up price matrix failed; expected message code not found in Spec!")
+		return nil
+	}
+	deliveryMsgIndex, ok := stream.Spec.GetCode(stream.RetrieveRequestMsg{})
+	if !ok {
+		log.Crit("setting up price matrix failed; expected message code not found in Spec!")
+		return nil
+	}
+
+	//assign the price tag to the message
+	priceMatrix[streamProtocol][uint64(retrieveRequestMsgIndex)] = &PriceTag{
+		Price:     big.NewInt(10),
+		SizeBased: false,
+		Direction: ChargeSender,
+	}
+
+	priceMatrix[streamProtocol][uint64(deliveryMsgIndex)] = &PriceTag{
+		Price:     big.NewInt(100),
+		SizeBased: true,
+		Direction: ChargeReceiver,
+	}
+
+	return priceMatrix
+
+}
+
+//Check if a message needs accounting
+func (dpo *DefaultPriceOracle) IsAccountedMsg(event *p2p.PeerEvent) bool {
+	var protoPriceMap map[uint64]*PriceTag
+	var ok bool
+	if protoPriceMap, ok = dpo.priceMatrix[event.Protocol]; !ok {
+		return false
+	}
+	if _, ok = protoPriceMap[*event.MsgCode]; !ok {
+		return false
+	}
+	return true
+}
+
+//Get the actual price of a message
+//If the message is size-based, it returns the calculated price
+func (dpo *DefaultPriceOracle) GetPriceForMsg(event *p2p.PeerEvent) (*big.Int, EntryDirection) {
+	if dpo.IsAccountedMsg(event) {
+		priceTag := dpo.priceMatrix[event.Protocol][*event.MsgCode]
+		price := &big.Int{}
+		price.Set(priceTag.Price)
+		if priceTag.SizeBased {
+			price.Mul(priceTag.Price, big.NewInt(int64(*event.MsgSize)))
+		}
+		return price, priceTag.Direction
+	}
+	return nil, false
+}
+
+//This swap implementation works by listening to message events on the p2p server.
+//It then handles the event received, filtering for messages and evaluating
+//if it needs accounting
+func (s *Swap) registerForEvents(srv *p2p.Server) {
+	go func() {
+		events := make(chan *p2p.PeerEvent)
+		sub := srv.SubscribeEvents(events)
+		defer sub.Unsubscribe()
+
+		for {
+			select {
+			case event := <-events:
+				go s.handleMsgEvent(event)
+			case err := <-sub.Err():
+				log.Error(err.Error())
+				return
+			}
+		}
+	}()
+}
+
+//Handle the message.
+//Determine if it needs accounting, and if yes, account for it
+func (s *Swap) handleMsgEvent(event *p2p.PeerEvent) {
+	if !s.priceOracle.IsAccountedMsg(event) {
+		return
+	}
+	s.AccountMsgForPeer(event)
+}
+
+//Do the accounting
+//Depending on the charging type of the message (set in the PriceTag),
+//it will charge the sender or receiver
+func (s *Swap) AccountMsgForPeer(event *p2p.PeerEvent) {
```

Comment:
should this be exported?

Reply:
I wonder myself - I reasoned that maybe there could be use cases where we may trigger accounting other than through the p2p event handling. 

But I guess in that case we can change it if we need, so probably it doesn't need to be exported now.